### ISSUE TITLE
Error with Login() function

### ISSUE THREAD

**Issue by gsoonh** [Soulter/hugging-chat-api#138]
I receive an error that points to login.py. I go thru' the code and realize it run "self.grantAuth(location)" and fail. What is location here?  The way we call the function is login(email, passwd). Where is this location if I may ask? Seeking your kind advice.

Here is the error code:
 
File "C:\Users\gsoonh\Documents\test\myenv\lib\site-packages\streamlit\runtime\scriptrunner\script_runner.py", line 534, in _run_script
    exec(code, module.__dict__)
File "C:\Users\gsoonh\demo.py", line 58, in <module>
    response = generate_response(prompt, hf_email, hf_pass)
File "C:\Users\gsoonh\demo.py", line 39, in generate_response
    cookies = sign.login()
File "C:\Users\gsoonh\Documents\test\myenv\lib\site-packages\hugchat\login.py", line 127, in login
    if self.grantAuth(location):
File "C:\Users\gsoonh\Documents\test\myenv\lib\site-packages\hugchat\login.py", line 105, in grantAuth
    raise Exception("grant auth fatal!")

**Comment by IA-Programming** [comment]
here's another error with the login function:

```python
---------------------------------------------------------------------------
MissingSchema                             Traceback (most recent call last)
<ipython-input-3-576e1a464eb7> in <cell line: 9>()
      7 # Log in to huggingface and grant authorization to huggingchat
      8 sign = Login(email, passwd)
----> 9 cookies = sign.login()

8 frames
/usr/local/lib/python3.10/dist-packages/hugchat/login.py in login(self)
    125         self.SigninWithEmail()
    126         location = self.getAuthURL()
--> 127         if self.grantAuth(location):
    128             return self.cookies
    129         else:

/usr/local/lib/python3.10/dist-packages/hugchat/login.py in grantAuth(self, url)
     98         if res.headers.__contains__("location"):
     99             location = res.headers["location"]
--> 100             res = self.requestsGet(location, allow_redirects=False)
    101             if res.cookies.__contains__("hf-chat"):
    102                 return 1

/usr/local/lib/python3.10/dist-packages/hugchat/login.py in requestsGet(self, url, params, allow_redirects)
     26 
     27     def requestsGet(self, url: str, params=None, allow_redirects=True) -> requests.Response:
---> 28         res = requests.get(
     29             url,
     30             params=params,

/usr/local/lib/python3.10/dist-packages/requests/api.py in get(url, params, **kwargs)
     71     """
     72 
---> 73     return request("get", url, params=params, **kwargs)
     74 
     75 

/usr/local/lib/python3.10/dist-packages/requests/api.py in request(method, url, **kwargs)
     57     # cases, and look like a memory leak in others.
     58     with sessions.Session() as session:
---> 59         return session.request(method=method, url=url, **kwargs)
     60 
     61 

/usr/local/lib/python3.10/dist-packages/requests/sessions.py in request(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)
    573             hooks=hooks,
    574         )
--> 575         prep = self.prepare_request(req)
    576 
    577         proxies = proxies or {}

/usr/local/lib/python3.10/dist-packages/requests/sessions.py in prepare_request(self, request)
    484 
    485         p = PreparedRequest()
--> 486         p.prepare(
    487             method=request.method.upper(),
    488             url=request.url,

/usr/local/lib/python3.10/dist-packages/requests/models.py in prepare(self, method, url, headers, files, data, params, auth, cookies, hooks, json)
    366 
    367         self.prepare_method(method)
--> 368         self.prepare_url(url, params)
    369         self.prepare_headers(headers)
    370         self.prepare_cookies(cookies)

/usr/local/lib/python3.10/dist-packages/requests/models.py in prepare_url(self, url, params)
    437 
    438         if not scheme:
--> 439             raise MissingSchema(
    440                 f"Invalid URL {url!r}: No scheme supplied. "
    441                 f"Perhaps you meant [https://{url}?](https://{url}/?)"

MissingSchema: Invalid URL '/login?next=%2Foauth%2Fauthorize%3Fclient_id%3D8f1a1d63-479b-46c8-84cb-521fe9f3222f%26scope%3Dopenid%2520profile%26response_type%3Dcode%26redirect_uri%3Dhttps%253A%252F%252Fhuggingface.co%252Fchat%252Flogin%252Fcallback%26state%3DeyJkYXRhIjp7ImV4cGlyYXRpb24iOjE3MDI1MjczMTU4OTcsInJlZGlyZWN0VXJsIjoiaHR0cHM6Ly9odWdnaW5nZmFjZS5jby9jaGF0L2xvZ2luL2NhbGxiYWNrIn0sInNpZ25hdHVyZSI6IjMzYzQyNmMwMTQ5ZGQ4ZTc4NzYzNWEwOWE5NTc1ZjYyZmVlZDI2MzgxOThmMTBmY2Q3MGNhYzAzY2E5MWU4ZmQifQ%253D%253D': No scheme supplied. Perhaps you meant [https:///login?next=%2Foauth%2Fauthorize%3Fclient_id%3D8f1a1d63-479b-46c8-84cb-521fe9f3222f%26scope%3Dopenid%2520profile%26response_type%3Dcode%26redirect_uri%3Dhttps%253A%252F%252Fhuggingface.co%252Fchat%252Flogin%252Fcallback%26state%3DeyJkYXRhIjp7ImV4cGlyYXRpb24iOjE3MDI1MjczMTU4OTcsInJlZGlyZWN0VXJsIjoiaHR0cHM6Ly9odWdnaW5nZmFjZS5jby9jaGF0L2xvZ2luL2NhbGxiYWNrIn0sInNpZ25hdHVyZSI6IjMzYzQyNmMwMTQ5ZGQ4ZTc4NzYzNWEwOWE5NTc1ZjYyZmVlZDI2MzgxOThmMTBmY2Q3MGNhYzAzY2E5MWU4ZmQifQ%253D%253D](https://login/?next=%2Foauth%2Fauthorize%3Fclient_id%3D8f1a1d63-479b-46c8-84cb-521fe9f3222f%26scope%3Dopenid%2520profile%26response_type%3Dcode%26redirect_uri%3Dhttps%253A%252F%252Fhuggingface.co%252Fchat%252Flogin%252Fcallback%26state%3DeyJkYXRhIjp7ImV4cGlyYXRpb24iOjE3MDI1MjczMTU4OTcsInJlZGlyZWN0VXJsIjoiaHR0cHM6Ly9odWdnaW5nZmFjZS5jby9jaGF0L2xvZ2luL2NhbGxiYWNrIn0sInNpZ25hdHVyZSI6IjMzYzQyNmMwMTQ5ZGQ4ZTc4NzYzNWEwOWE5NTc1ZjYyZmVlZDI2MzgxOThmMTBmY2Q3MGNhYzAzY2E5MWU4ZmQifQ%253D%253D)?
```

It seems that is detecting or extracting a relative link

**Comment by Whitelisted1** [comment]
Unfortunately, I am unable to reproduce this error, although I will try to help the best I can.
From the information in the logs provided by @IA-Programming, it looks like `grantAuth` is extracting a relative URL from the headers of a response from the server.
If possible, try adding the below code to the `grantAuth` function in `login.py` on your end.

```python
    def grantAuth(self, url: str) -> int:
        res = self.requestsGet(url, allow_redirects=False)
        if res.headers.__contains__("location"):
            location = res.headers["location"]

            # Add the below code
            logging.debug(location)
            if location[0] == "/":
                location = "https://huggingface.co/chat" + location
            logging.debug(location)
            # Add the above code

            res = self.requestsGet(location, allow_redirects=False)
            if res.cookies.__contains__("hf-chat"):
                return 1
        # ...
```

Along with that, make sure you enable logging in your own script
```python
import logging
logging.basicConfig(level=logging.DEBUG)
```

Again, I apologize for my inability to reproduce the error and hope that this helps with your issues. Once you have tested the code, please let me know the result.

**Comment by IA-Programming** [comment]
here in the streamlit demo of this application the error can be seen briefly, but I will try it out the solution that you sugest @Whitelisted1 :
[Free personal AI Assistant with plugins](https://github.com/IntelligenzaArtificiale/Free-personal-AI-Assistant-with-plugin)

**Comment by Whitelisted1** [comment]
I wanted to dig into the issue more and was able to reproduce the error by repeatedly logging in. Using the code I provided (with some more modification), I was able to produce a valid url.
The URL that was provided did have an issue, though. Instead of providing the script with the `hf-chat` authentication, it seems to have just redirected to [HuggingFace's login page](https://huggingface.co/login).
